File: 2021/week_11/210915/9663.cpp
#include <stdio.h>
#include <math.h>
#define NUM 16

int dfs(int queen[], int n, int row){
    int count = 0;
    // n개의 행에 퀸들이 있다면 종료
    if(n == row) {
        return 1;
    }

    for(int col = 0; col < n; col++) {
        queen[row] = col;
        int is_possible = 1;
        for(int x = 0; x < row; x++) {
            if (queen[x] == queen[row]) {
                is_possible = 0;
                break;
            }
            if (abs(queen[x] - queen[row]) == (row - x)) {
                is_possible = 0;
                break;
            }
        }
        if (is_possible) count += dfs(queen, n, row + 1);
    }
    return count;
}

int main(void){
    int n; scanf("%d", &n);
    int queen[NUM];
    printf("%d", dfs(queen, n, 0));
}

File: 2021/week_11/210915/9663.py
def dfs(queens, n, row):
    count = 0
    # n개의 행에 퀸들이 있다면 종료 
    if n == row:
        return 1
    # 모든 열 탐색
    for col in range(n):
        queens[row] = col
        for x in range(row):
            # 같은 열에 있다면
            if queens[x] == queens[row]:
                break
            # 대각선에 있다면
            if abs(queens[x] - queens[row]) == (row - x):
                break
        else:
            count += dfs(queens, n, row + 1)
    
    return count

if __name__ == "__main__":
    n = int(input())
    print(dfs([0] * n,  n, 0))
